acme/src/lib.rs

```diff
@@ -0,0 +1,149 @@
+//! configured by `trillium-acme` will respond to.
+//!
+//! You must supply a cache via [`AcmeConfig::cache`] or one of the other cache methods. This cache
+//! will keep the ACME account key and registered certificates between runs, needed to avoid
+//! hitting rate limits. You can use [`rustls_acme::caches::DirCache`] for a simple filesystem
+//! cache, or implement your own caching using the `rustls_acme` cache traits.
+//!
+//! By default, `trillium-acme` will use the Let's Encrypt staging environment, which is suitable
+//! for testing purposes; it produces certificates signed by a staging root so that you can verify
+//! your stack is working, but those certificates will not be trusted in browsers or other HTTPS
+//! clients. The staging environment has more generous rate limits for use while testing.
+//!
+//! When you're ready to deploy to production, you can call `.directory_lets_encrypt(true)` to
+//! switch to the production Let's Encrypt environment, which produces certificates trusted in
+//! browsers and other HTTPS clients. The production environment has
+//! [stricter rate limits](https://letsencrypt.org/docs/rate-limits/).
+//!
+//! `trillium-acme` builds upon the [`rustls-acme`](https://crates.io/crates/rustls-acme) crate.
+
+#![forbid(unsafe_code)]
+#![deny(
+    clippy::dbg_macro,
+    missing_copy_implementations,
+    rustdoc::missing_crate_level_docs,
+    missing_debug_implementations,
+    missing_docs,
+    nonstandard_style,
+    unused_qualifications
+)]
+
+use std::fmt::Debug;
+use std::future::Future;
+use std::sync::Arc;
+
+use futures_lite::StreamExt;
+use rustls_acme::futures_rustls::rustls::ServerConfig;
+use rustls_acme::AcmeAcceptor;
+use trillium::log::{error, info};
+use trillium_server_common::async_trait;
+
+pub use rustls_acme::{self, AcmeConfig};
+
+mod transport;
+pub use transport::Transport;
+
+/// An acceptor that handles ACME tls-alpn-01 challenges.
+///
+/// After processing a challenge, this acceptor will return a Transport representing a closed
+/// connection.
+#[derive(Clone)]
+pub struct Acceptor {
+    acme_acceptor: AcmeAcceptor,
+    server_config: Arc<ServerConfig>,
+}
+
+impl Debug for Acceptor {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Acceptor").finish_non_exhaustive()
+    }
+}
+
+/// Create a new [`Acceptor`] to pass to [`trillium_server_common::Config::with_acceptor`], and a
+/// new future that must be spawned detached in the background.
+pub fn new<EC: 'static + Debug, EA: 'static + Debug>(
+    config: AcmeConfig<EC, EA>,
+) -> (Acceptor, impl Future) {
+    let mut state = config.state();
+    let server_config = Arc::new(
+        ServerConfig::builder()
+            .with_safe_defaults()
+            .with_no_client_auth()
+            .with_cert_resolver(state.resolver()),
+    );
+    let acme_acceptor = state.acceptor();
+
+    let future = async move {
```

Comment:
This future is awkward, and possibly an indication that trillium needs a new affordance to spawn tasks from within init. When does this stream receive items? Is it generally sequenced with the acme_acceptor returning None? Is it fully async and unrelated to the request/response cycle?

Reply:
It's fully async, and handles things like renewal.

I would be happy to hide this detail, but 1) I would need a runtime-independent way to spawn a task, and 2) I would need access to the stopper, which also needs to be runtime-independent.